File: awk_tool.py
import re
from pathlib import Path
from check_file_permission import check_file_permission

def awk_file(file_path, pattern, action, config):
    if not check_file_permission(file_path, "write", config):
        raise PermissionError(f"Write access denied for {file_path}")
    file_path = Path(file_path).resolve()
    if not file_path.is_file():
        raise ValueError(f"File not found: {file_path}")
    result = []
    with open(file_path, "r", encoding="utf-8") as f:
        for line in f:
            if re.search(pattern, line):
                result.append(action(line.strip()))
            else:
                result.append(line.strip())
    with open(file_path, "w", encoding="utf-8") as f:
        f.write("\n".join(result))
    return f"Processed {file_path} with awk"

File: ts-tools/awk_tool.ts
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import * as fs from 'fs/promises';
import * as path from 'path';
import { checkFilePermission } from './check_file_permission';

export async function awkFile(
  filePath: string,
  pattern: string,
  action: (line: string) => string,
  config: any,
): Promise<string> {
  await checkFilePermission(filePath, 'write', config);
  const resolvedPath = path.resolve(filePath);
  const data = await fs.readFile(resolvedPath, 'utf-8');
  const lines = data.split('\n');
  const result: string[] = [];
  const regex = new RegExp(pattern);
  for (const line of lines) {
    if (regex.test(line)) {
      result.push(action(line));
    } else {
      result.push(line);
    }
  }
  await fs.writeFile(resolvedPath, result.join('\n'), 'utf-8');
  return `Processed ${filePath} with awk`;
}
